### PR TITLE
README for hacking CLI commands

### PR DESCRIPTION
**Pull request by fabianofranz**
I started a draft of a guide for CLI contributors. WIP, very early version, FYI @smarterclayton @liggitt @deads2k

### CODE REVIEW

**Comment by liggitt** [comment]
+1. Mention the scripts to run to update generated conversions and docs

**Comment by fabianofranz** [comment]
@liggitt @smarterclayton I know there are things missing, but I think this is a good starting point. Please review.

**Comment by smarterclayton** [comment]
Belongs in the docs dir

File: cmd/README.adoc [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+for my command.`
+
+  mineExample = `  // Run my command's first action
+  $ %[1]s first
+
+  // Run my command's second action on latest stuff
+  $ %[1]s second --latest`
+)
+
+func NewCmdMine(fullName string, f *clientcmd.Factory, out io.Writer) *cobra.Command {
```
Comment (deads2k) [inline cmd/README.adoc]: `name, fullName string`

File: cmd/README.adoc [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+  deployLatest bool
+  retryDeploy  bool
+}
+
+const (
+  deployLong = `Some long description 
+for the deploy command.` // <2>
+
+  deployExample = `  // Display the latest deployment for the 'database' DeploymentConfig
+  $ %[1]s deploy database
```
Comment (deads2k) [inline cmd/README.adoc]: use an example with `fullName` (`oc deploy`) here.

File: cmd/README.adoc [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+
+  // Start a new deployment based on the 'database' DeploymentConfig
+  $ %[1]s deploy database --latest` // <3>
+)
+
+func NewCmdDeploy(fullName string, f *clientcmd.Factory, out io.Writer) *cobra.Command { // <4>
+  options := &DeployOptions{}
+
+  cmd := &cobra.Command{
+    Use:     "deploy DEPLOYMENTCONFIG", <5>
```
Comment (deads2k) [inline cmd/README.adoc]: `name`

**Comment by fabianofranz** [comment]
@deads2k comments addressed, thanks.
@smarterclayton moved, actually I merged it with the [existing cli doc](https://github.com/openshift/origin/blob/master/docs/cli.md). The examples were replaced by a link to the "by example" docs that are auto-generated so they don't get outdated. The section about common flags and output formats were kept.

**Comment by ironcladlou** [comment]
Why not formalize the command structure it with an actual interface?

**Comment by fabianofranz** [comment]
@ironcladlou Yes, I'm thinking about it. @kargakis also suggested a format structure.

**Comment by kargakis** [comment]
> @smarterclayton moved, actually I merged it with the existing cli doc.

@fabianofranz this is more sort of "Hacking the cli" while cli.md is about usage. There is an existing [PR](https://github.com/openshift/origin/pull/3559) updating it. We should keep both.

**Comment by fabianofranz** [comment]
@kargakis Makes sense, I merged mostly because of deprecation concerns but it's obvious now that cli.md is active. Let me split them again.

**Comment by fabianofranz** [comment]
Ok, fixed. `cli.md` is back and now we have `cli_hacking_guide.adoc` as a document more oriented to actually *contributing* to CLI commands. CC @kargakis @smarterclayton

**Comment by fabianofranz** [comment]
Other comments here, folks?

**Comment by deads2k** [comment]
> @ironcladlou Yes, I'm thinking about it. @kargakis also suggested a formal structure.

I don't know that I'd want to force a one size fits all model.  I'm fine with slightly different method signatures based on command need, but I think the general approach makes it easier for people to review and maintain the commands.

**Comment by deads2k** [comment]
> Other comments here, folks?

lgtm

File: docs/cli_hacking_guide.adoc [hunk 4]
```diff
@@ -0,0 +1,293 @@
+= OpenShift 3 Command Line Interface (CLI)
```
Comment (kargakis) [inline docs/cli_hacking_guide.adoc]: Both of our cli docs have the same header, should we change this to **CLI Hacking Guide** and remove the introduction about the cli (we already introduce it in cli.md) ?

Comment (fabianofranz) [inline docs/cli_hacking_guide.adoc]: Agreed, header changed. I kept the introduction for people that reach this doc first, but added links to cli.md.

Comment (kargakis) [inline docs/cli_hacking_guide.adoc]: Sup

**Comment by kargakis** [comment]
LGTM

**Comment by fabianofranz** [comment]
[merge]

**Pull request merged**